File: hive/src/main/java/io/delta/hive/DeltaRecordReaderWrapper.java
package io.delta.hive;

import com.google.common.base.Joiner;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import scala.collection.JavaConverters;

import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hive.ql.io.parquet.read.ParquetRecordReaderWrapper;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorConverters;
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory;
import org.apache.hadoop.io.ArrayWritable;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.io.Writable;
import org.apache.hadoop.mapred.FileSplit;
import org.apache.hadoop.mapred.InputSplit;
import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.Reporter;
import org.apache.parquet.hadoop.ParquetInputFormat;
import org.apache.spark.sql.delta.DeltaHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DeltaRecordReaderWrapper extends ParquetRecordReaderWrapper {

    private static final Logger LOG = LoggerFactory.getLogger(DeltaRecordReaderWrapper.class);

    private Writable[] partitionWritable = null;

    public DeltaRecordReaderWrapper(ParquetInputFormat<ArrayWritable> newInputFormat, InputSplit oldSplit, JobConf oldJobConf, Reporter reporter) throws IOException, InterruptedException {
        super(newInputFormat, oldSplit, oldJobConf, reporter);

        if (!(oldSplit instanceof FileSplit)) {
            throw new IllegalArgumentException("Unknown split type: " + oldSplit);
        } else {
            final String columnNameProperty = oldJobConf.get(DeltaStorageHandler.DELTA_PARTITION_COLS_NAMES);
            final String columnTypeProperty = oldJobConf.get(DeltaStorageHandler.DELTA_PARTITION_COLS_TYPES);
            LOG.info("Delta partition cols: " + columnNameProperty + " with types: " + columnTypeProperty);

            if (columnNameProperty == null || columnNameProperty.trim().length() == 0
                    || columnTypeProperty == null || columnTypeProperty.trim().length() == 0) {
                LOG.info("No partition info is provided...");
            } else {
                // generate partition writale values which will be appended after data values from parquet
                final List<String> columnNames;
                final List<String> columnTypes;
                if (columnNameProperty.length() == 0) {
                    columnNames = new ArrayList<String>();
                } else {
                    columnNames = Arrays.asList(columnNameProperty.split(","));
                }
                if (columnTypeProperty.length() == 0) {
                    columnTypes = new ArrayList<String>();
                } else {
                    columnTypes = Arrays.asList(columnTypeProperty.split(":"));
                }

                Path filePath = ((FileSplit)oldSplit).getPath();

                Map<String, String> parsedPartitions = JavaConverters.mapAsJavaMap(
                        DeltaHelper.parsePathPartition(filePath, JavaConverters.asScalaBufferConverter(columnNames).asScala().toSeq()));

                partitionWritable = new Writable[columnNames.size()];
                // inspect partition values
                for(int i=0; i < columnNames.size(); i++) {
                    ObjectInspector oi = PrimitiveObjectInspectorFactory
                            .getPrimitiveWritableObjectInspector(TypeInfoFactory
                                    .getPrimitiveTypeInfo(columnTypes.get(i)));

                    partitionWritable[i] = (Writable)ObjectInspectorConverters.getConverter(
                            PrimitiveObjectInspectorFactory.javaStringObjectInspector, oi).convert(parsedPartitions.get(columnNames.get(i)));
                }
                LOG.info("Parsed partition values from " + filePath.toString() + " list: " + Joiner.on(",").withKeyValueSeparator("=").join(parsedPartitions)
                    + ", partitionWritable length:" + partitionWritable.length);
            }
        }
    }

    @Override
    public boolean next(NullWritable key, ArrayWritable value) throws IOException {
        boolean hasNext = super.next(key, value);
        if (partitionWritable != null && partitionWritable.length != 0) {
            // append partition values to data values
            for (int i=0; i < partitionWritable.length; i++) {
                value.get()[value.get().length - partitionWritable.length + i] = partitionWritable[i];
            }
        }
        return hasNext;
    }

    @Override
    public ArrayWritable createValue() {
        ArrayWritable value = super.createValue();
        if (partitionWritable != null && partitionWritable.length != 0) {
            for (int i=0; i < partitionWritable.length; i++) {
                value.get()[value.get().length - partitionWritable.length + i] = partitionWritable[i];
            }
        }
        return value;
    }
}

File: hive/src/main/scala/io/delta/hive/DeltaRecordReaderWrapper.scala
/*
 * Copyright 2019 Databricks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.hive

import org.apache.hadoop.hive.ql.io.parquet.read.ParquetRecordReaderWrapper
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorConverters
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory
import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory
import org.apache.hadoop.io.ArrayWritable
import org.apache.hadoop.io.NullWritable
import org.apache.hadoop.io.Writable
import org.apache.hadoop.mapred.JobConf
import org.apache.hadoop.mapred.Reporter
import org.apache.parquet.hadoop.ParquetInputFormat
import org.slf4j.LoggerFactory

/**
 * A record reader that reads data from the underlying Parquet reader and inserts partition values
 * which don't exist in the Parquet files.
 *
 * As we have verified the Hive schema in metastore is consistent with the Delta schema, the row
 * returned by the underlying Parquet reader will match the Delta schema except that it leaves all
 * partition columns as `null` since they are not in the raw parquet files. Hence, for the missing
 * partition values, we need to use the partition information in [[DeltaInputSplit]] to create the
 * corresponding [[Writable]]s, and insert them into the corresponding positions when reading a row.
 */
class DeltaRecordReaderWrapper(
    inputFormat: ParquetInputFormat[ArrayWritable],
    split: DeltaInputSplit,
    jobConf: JobConf,
    reporter: Reporter) extends ParquetRecordReaderWrapper(inputFormat, split, jobConf, reporter) {

  private val LOG = LoggerFactory.getLogger(classOf[DeltaRecordReaderWrapper])

  /** The positions of partition columns in Delta schema and their corresponding values. */
  private val partitionValues: Array[(Int, Writable)] =
    split.getPartitionColumns.map { partition =>
      val oi = PrimitiveObjectInspectorFactory
        .getPrimitiveWritableObjectInspector(TypeInfoFactory
          .getPrimitiveTypeInfo(partition.tpe))
      val partitionValue = ObjectInspectorConverters.getConverter(
        PrimitiveObjectInspectorFactory.javaStringObjectInspector,
        oi).convert(partition.value).asInstanceOf[Writable]
      (partition.index, partitionValue)
    }

  override def next(key: NullWritable, value: ArrayWritable): Boolean = {
    val hasNext = super.next(key, value)
    // TODO Figure out when the parent reader resets partition columns to null so that we may come
    // out a better solution to not insert partition values for each row.
    if (hasNext) {
      insertPartitionValues(value)
    }
    hasNext
  }

  /**
   * As partition columns are not in the parquet files, they will be set to `null`s every time
   * `next` is called. We should insert partition values manually for each row.
   */
  private def insertPartitionValues(value: ArrayWritable): Unit = {
    val valueArray = value.get()
    var i = 0
    val n = partitionValues.length
    // Using while loop for better performance since this method is called for each row.
    while (i < n) {
      val partition = partitionValues(i)
      // The schema of `valueArray` is the Hive schema, and it's the same as the Delta
      // schema since we have verified it in `DeltaInputFormat`. Hence, the position of a partition
      // column in `valueArray` is the same as its position in Delta schema.
      valueArray(partition._1) = partition._2
      i += 1
    }
  }
}
